### ISSUE TITLE
HeadOutlet not working when using WebAssembly and Serverside Pre Render.

### ISSUE THREAD

**Issue by Bartmax** [dotnet/aspnetcore#37254]
### Describe the bug
Using HeadOutlet / PageTitle component to change the page title in  WebAssembly *with* Serverside-Prerender, theres no way to get the title prerendered. 
First page render always renders the server-side title (or none if title is not specified at serverside).
If title is specified, title is also _not_ updated after webassembly boots.

### Expected behavior
We should get server-side `<title>` tag when doing serverside prerender that reflects the page of the current url before any client-side code is executed.
This is  *very importante* for SEO, social sharing excerpts and title prerender.

### To Reproduce
`_Host.cshtml`:
```C#
<head>
<title>Host: Server</title> 
...c#
 <component type="typeof(HeadOutlet)" render-mode="WebAssemblyPrerendered" />
</head>
``` 
_note_ on `_Host.cshtml` `<title>` tag: If we do not include the `<title>...</title>` in the `<head>` on html source, 
after webassembly boots the client side updates the page title, and we get no prerendered title.
If we keep the `<title>` tag on the `_Host.cshtml`, we get this "generic" title from serverside and the 
first page never reflects/update to the webassembly `<PageTitle>` component.
<hr>

`Index.blazor`:
```c#
@page "/"
<PageTitle>Main page: Blazor</PageTitle>
```

`SomePage.blazor`:
```c#
@page "/somepage"
<PageTitle>Some Page</PageTitle>
``` 

On first render, the title will show the server one: "Host: Server"
when navigating normally the title will change just fine.
If we refresh (F5, CTRL+R) at `/somepage` the title will display: "Host: Server" again.

If I change `_Host.cshtml` to:
```c#
<head>
...
 <component type="typeof(HeadOutlet)" render-mode="ServerPrerendered" />
</head>
```
a client-side error is returned:
```
Error: Found malformed component comment at Blazor:{"prerenderId":"3642bc95455b4cfa882b1fa6b3204edf"}
```

### Repro:
I included a repo for your convenience at: https://github.com/Bartmax/HeadOutletBug

### Further technical details
```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-rc.1.21463.6
 Commit:    e627d556a1
```
<details>
<summary>
see more technical info
</summary>

```
Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22468
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-rc.1.21463.6\

Host (useful for support):
  Version: 6.0.0-rc.1.21451.13
  Commit:  d7619cd4b1
```

</details>

**Comment by javiercn** [comment]
@Bartmax thanks for contacting us.

@SteveSandersonMS we fix something around this area, didn't we?

**Comment by SteveSandersonMS** [comment]
Yes, we even have an E2E test case at `src\Components\WebAssembly\testassets\Wasm.Prerendered.Server` that shows this working.

@Bartmax Since we can see this working in the current version of our sources, could you please retry with an RC2 or RTM-branded daily build of the SDK from https://github.com/dotnet/installer/blob/main/README.md#installers-and-binaries? Please let us know if you're still unable to make it work. Also consider checking the example in our sources at `src\Components\WebAssembly\testassets\Wasm.Prerendered.Server`.

**Comment by Bartmax** [comment]
Updated to RTM the difference is that now the server-side `<title>` tag, when exists, it gets updated correctly, this is better but still I believe is far from ideal.

Why?
Because I will still have only 1 server-side title for the whole site. Meaning that sharing/seo/etc on any page will show that same title (and I believe other head elements, but I didn't try) for all pages.

So for example having these 3 endpoints:

site.com/products/nike
site.com/event-woman-10k
site.com/

If I enter any of those links, all three will render the same title server-side wise. (aka view source). I believe this is unavoidable when not using prerender but it's one of the main reasons one would want to pre-render.

There are others meta and link tags that are very important to be page specific, off the top of my head: canonical, preload, description, and title to name a few.

I totally believe that this should reads: Counter _when doing serverside_ prerendering.
![image](https://user-images.githubusercontent.com/3386797/135898939-05525251-4b70-4ba8-93e2-613d9e4d71c5.png)

**Comment by Bartmax** [comment]
I confirm, head elements works the same.
When using
```
<HeadContent>
    <meta description="counts increments when a button is clicked" />
</HeadContent>
```

the meta is added via client-side (it shows in inspector) but not at server-side. (it doesn't show on source)

![image](https://user-images.githubusercontent.com/3386797/135900153-590a7aa8-defb-4038-bb64-ad6a35455262.png)

**Comment by Bartmax** [comment]
I also confirm that on Blazor Server it works perfect
![image](https://user-images.githubusercontent.com/3386797/135903404-39f3e4a4-41d2-459a-8587-fe156655d793.png)

**Comment by Bartmax** [comment]
Ok, something is definitively going on and I think I found the bug.
if I author this `_Host.cshtml`:
```c#
<div id="app">
<component type="typeof(HeadOutlet)" render-mode="WebAssemblyPrerendered" /> 
<component type="typeof(HeadOutletBug.Client.App)" render-mode="WebAssemblyPrerendered" />
</div>
```
The `<title>` tag is not rendered.

BUT if I alter the order (the `App` before the `HeadOutlet`)
```c#
<div id="app">
<component type="typeof(HeadOutletBug.Client.App)" render-mode="WebAssemblyPrerendered" />
<component type="typeof(HeadOutlet)" render-mode="WebAssemblyPrerendered" /> 
</div>
```

the `<title>` tag gets rendered, but of course, not in the right place (inside the `<head>`).

**Comment by SteveSandersonMS** [comment]
As far as I'm aware, this does work correctly with the default project template for a Blazor Server application. Are you able to check whether the default project does work as you expect, and compare your application against that?

**Comment by Bartmax** [comment]
Blazor Server work as expected. This error is on WebAssembly with AspNet Core Hosted with Pre Render. (there is no template for that scenario)

**Comment by Bartmax** [comment]
You can take a look at this [commit](https://github.com/Bartmax/HeadOutletBug/commit/8f6dac94714870625089207157ec38177304e183) to see the updates to WebAssembly default template. This commit adds server pre-render to the project.

**Comment by Bartmax** [comment]
Ok, I guess I got it thanks to your description of the issue here https://github.com/dotnet/aspnetcore/issues/37293

So the workaround is tu put the content of the _Host in a _Layout. I'll try that.
